File: python/mirage/shell/console/pipe.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import asyncio
from collections import deque
from collections.abc import AsyncIterator

from mirage.shell.console.job_console import JobConsole
from mirage.shell.console.types import Channel
from mirage.shell.errors import PipeClosed


class PipeConsole(JobConsole):
    """A single-reader pipe whose writes follow consumer demand.

    A delivered chunk is acknowledged only when the reader advances. Closing
    after one chunk therefore wakes the writer without fetching another
    backend page merely to discover that the consumer has already stopped.
    """

    def __init__(self, pipe_stderr: bool = False) -> None:
        super().__init__()
        self._pipe_stderr = pipe_stderr
        self._chunks: deque[bytes] = deque()
        self._bytes = 0
        self._ended = False
        self._delivered = 0
        self._accepted = 0
        self._reader_closed = False
        self._failure: BaseException | None = None
        self._changed = asyncio.Event()

    async def emit(self, channel: Channel, data: bytes) -> None:
        if channel != Channel.STDOUT and not self._pipe_stderr:
            await super().emit(channel, data)
            return
        if not data:
            return
        while self._bytes >= 65536 and not self._reader_closed:
            self._changed.clear()
            await self._changed.wait()
        if self._reader_closed:
            raise PipeClosed()
        self._chunks.append(data)
        self._bytes += len(data)
        self._delivered += 1
        ticket = self._delivered
        self._changed.set()
        while self._accepted < ticket and not self._reader_closed:
            self._changed.clear()
            await self._changed.wait()

    @property
    def closed_reader(self) -> bool:
        return self._reader_closed

    def end(self, error: BaseException | None = None) -> None:
        if error is not None:
            self._failure = error
        self._ended = True
        self._changed.set()

    def close_reader(self) -> None:
        self._reader_closed = True
        self._chunks.clear()
        self._bytes = 0
        self._changed.set()

    async def stream(self) -> AsyncIterator[bytes]:
        try:
            while True:
                if self._chunks:
                    chunk = self._chunks.popleft()
                    self._bytes -= len(chunk)
                    self._changed.set()
                    yield chunk
                    self._accepted += 1
                    self._changed.set()
                elif self._ended:
                    if self._failure is not None:
                        raise self._failure
                    return
                else:
                    self._changed.clear()
                    await self._changed.wait()
        finally:
            self.close_reader()

File: typescript/packages/core/src/shell/console/pipe.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { JobConsole } from './job_console.ts'
import { Channel } from './types.ts'
import { PipeClosed } from '../errors.ts'

/** A single-reader pipe with a kernel-sized buffer. A write succeeds while
 * the reader is open and the buffer has room, as in bash, where a producer
 * that finishes before `head` closes keeps its status. A lazy source is
 * pulled only after `drain`, so closing after one chunk does not fetch
 * another backend page first. */
export class PipeConsole extends JobConsole {
  private chunks: Uint8Array[] = []
  private bytes = 0
  private ended = false
  private readerClosed = false
  private delivered = 0
  private accepted = 0
  private failure: Error | undefined
  private wake: (() => void)[] = []

  constructor(private readonly pipeStderr = false) {
    super()
  }

  private notify(): void {
    for (const resolve of this.wake.splice(0)) resolve()
  }

  private changed(): Promise<void> {
    return new Promise((resolve) => this.wake.push(resolve))
  }

  override async emit(channel: Channel, data: Uint8Array): Promise<void> {
    if (channel !== Channel.STDOUT && !this.pipeStderr) {
      await super.emit(channel, data)
      return
    }
    if (data.byteLength === 0) return
    while (this.bytes >= 65536 && !this.readerClosed) await this.changed()
    if (this.readerClosed) throw new PipeClosed()
    this.chunks.push(data)
    this.bytes += data.byteLength
    this.delivered += 1
    this.notify()
  }

  /** Wait until the reader has taken every chunk or closed. */
  async drain(): Promise<void> {
    while (this.accepted < this.delivered && !this.readerClosed) await this.changed()
  }

  get closedReader(): boolean {
    return this.readerClosed
  }

  end(error?: unknown): void {
    if (error !== undefined)
      this.failure =
        error instanceof Error ? error : new Error('Pipeline producer failed', { cause: error })
    this.ended = true
    this.notify()
  }

  closeReader(): void {
    this.readerClosed = true
    this.chunks = []
    this.bytes = 0
    this.notify()
  }

  /** The reader is done. The close lands on the next task-queue turn, the
   * way a kernel pipe closes when its reader exits: a writer that is not
   * blocked finishes its burst into the buffer first, as it does on the
   * python host, whose reader runs only once the writer suspends. */
  release(): void {
    setTimeout(() => {
      this.closeReader()
    }, 0)
  }

  async *stream(): AsyncGenerator<Uint8Array> {
    try {
      for (;;) {
        const chunk = this.chunks.shift()
        if (chunk !== undefined) {
          this.bytes -= chunk.byteLength
          this.notify()
          yield chunk
          this.accepted += 1
          this.notify()
        } else if (this.ended) {
          if (this.failure !== undefined) throw this.failure
          return
        } else {
          await this.changed()
        }
      }
    } finally {
      this.release()
    }
  }
}
